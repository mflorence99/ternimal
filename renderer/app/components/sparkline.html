<figure #wrapper class="sparkline">
  <canvas #canvas> </canvas>
</figure>
