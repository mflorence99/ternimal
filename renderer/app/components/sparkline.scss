:host {
  display: block;
}

.sparkline {
  height: 100%;
  width: 100%;
}
