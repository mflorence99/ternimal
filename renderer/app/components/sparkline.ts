import { Params } from '../services/params';
import { Utils } from '../services/utils';

import { ChangeDetectionStrategy } from '@angular/core';
import { Component } from '@angular/core';
import { ElementRef } from '@angular/core';
import { Input } from '@angular/core';
import { OnInit } from '@angular/core';
import { ViewChild } from '@angular/core';

import Chart from 'chart.js';

interface Sparkline {
  data: number[];
  labels: string[];
}

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'ternimal-sparkline',
  templateUrl: 'sparkline.html',
  styleUrls: ['sparkline.scss']
})
export class SparklineComponent implements OnInit {
  /* eslint-disable @typescript-eslint/member-ordering */

  chart: Chart;
  @Input()
  get sparkline(): Sparkline {
    return this._sparkline;
  }
  set sparkline(data: Sparkline) {
    this._sparkline = data;
    this.updateChart();
  }

  @ViewChild('canvas', { static: true }) canvas: ElementRef;
  @ViewChild('wrapper', { static: true }) wrapper: ElementRef;

  private _sparkline: Sparkline;

  constructor(
    private host: ElementRef,
    private params: Params,
    private utils: Utils
  ) {}

  ngOnInit(): void {
    const canvas = this.canvas.nativeElement;
    canvas.height = this.wrapper.nativeElement.offsetHeight;
    canvas.width = this.wrapper.nativeElement.offsetWidth;
    const ctx = canvas.getContext('2d');
    this.chart = new Chart(ctx, {
      type: 'line',
      data: {
        labels: [],
        datasets: [
          {
            backgroundColor: this.utils.colorOf(
              this.host,
              this.params.rgb.green,
              0.25
            ),
            borderColor: this.utils.colorOf(
              this.host,
              this.params.rgb.green,
              1
            ),
            data: []
          },
          {
            backgroundColor: this.utils.colorOf(
              this.host,
              this.params.rgb.yellow,
              0.25
            ),
            borderColor: this.utils.colorOf(
              this.host,
              this.params.rgb.yellow,
              1
            ),
            data: []
          },
          {
            backgroundColor: this.utils.colorOf(
              this.host,
              this.params.rgb.red,
              0.25
            ),
            borderColor: this.utils.colorOf(this.host, this.params.rgb.red, 1),
            data: []
          }
        ]
      },
      options: {
        responsive: false,
        legend: {
          display: false
        },
        elements: {
          line: {
            borderWidth: 1
          },
          point: {
            radius: 0
          }
        },
        tooltips: {
          enabled: false
        },
        scales: {
          yAxes: [
            {
              display: false,
              ticks: { suggestedMin: 0, suggestedMax: 100 }
            }
          ],
          xAxes: [
            {
              display: false
            }
          ]
        }
      }
    });
    // set the initial data
    this.updateChart();
  }

  // private methods

  private updateChart(): void {
    if (this.chart && this.sparkline) {
      this.chart.data.labels = this.sparkline.labels;
      const green = new Array(this.sparkline.data.length).fill(NaN);
      const yellow = new Array(this.sparkline.data.length).fill(NaN);
      const red = new Array(this.sparkline.data.length).fill(NaN);
      this.sparkline.data.forEach((value, ix) => {
        if (value > 66) red[ix] = value;
        else if (value > 33) yellow[ix] = value;
        else green[ix] = value;
      });
      this.chart.data.datasets[0].data = green;
      this.chart.data.datasets[1].data = yellow;
      this.chart.data.datasets[2].data = red;
      this.chart.update();
    }
  }
}
