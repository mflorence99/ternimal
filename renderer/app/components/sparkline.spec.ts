import { SparklineComponent } from './sparkline';

import { prepare } from './component.spec';

import 'jest-extended';

import { ComponentFixture } from '@angular/core/testing';
import { TestBed } from '@angular/core/testing';

describe('SparklineComponent', () => {
  let component: SparklineComponent;
  let fixture: ComponentFixture<SparklineComponent>;

  beforeEach(() => {
    prepare();
    fixture = TestBed.createComponent(SparklineComponent);
    component = fixture.componentInstance;
  });

  test('sparkline accessor / ngOnInit', () => {
    const data = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100];
    const labels = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10'];
    component.sparkline = { data, labels };
    component.ngOnInit();
    expect(component.chart.data.labels).toEqual(labels);
    // segments divided into red/yellow/green depending on thresholds
    const green = [10, 20, 30, NaN, NaN, NaN, NaN, NaN, NaN, NaN];
    const yellow = [NaN, NaN, NaN, 40, 50, 60, NaN, NaN, NaN, NaN];
    const red = [NaN, NaN, NaN, NaN, NaN, NaN, 70, 80, 90, 100];
    expect(component.chart.data.datasets[0].data).toEqual(green);
    expect(component.chart.data.datasets[1].data).toEqual(yellow);
    expect(component.chart.data.datasets[2].data).toEqual(red);
  });

  test('snapshot', () => {
    const data = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100];
    const labels = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10'];
    component.sparkline = { data, labels };
    component.ngOnInit();
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
