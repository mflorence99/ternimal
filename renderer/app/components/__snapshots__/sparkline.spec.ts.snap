// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SparklineComponent snapshot 1`] = `
<ternimal-sparkline
  __ngContext__={[Function LRootView]}
  _sparkline={[Function Object]}
  canvas={[Function ElementRef]}
  chart={[Function Chart]}
  host={[Function ElementRef]}
  params={[Function Params]}
  utils={[Function Utils]}
  wrapper={[Function ElementRef]}
>
  <figure
    class="sparkline"
  >
    <canvas
      height="0"
      style="display: block;"
      width="0"
    />
  </figure>
</ternimal-sparkline>
`;
